### PR TITLE
Xref backend for beancount-mode

### PR DESCRIPTION
**Pull request by vkazanov**
Hello again, 

I've added another quality-of-life improvement I felt was both useful and easy enough to implement: an Xref backend. This adds support for:

1. Jumping up account definitions (aka open directive) under point using the default =M-.= binding
2. finding account under point references using =M-?=.
3. grepping through known symbol references and definitions using apropos like search with =C-M-.=.

While it is not hard to use basic tools such as grep with Beancount files, I felt that working from standard tooling aligns well with other modes supporting Xref and would be helpful to people used to default keybindings. 

I checked it in Emacs 26, 27, 28, 29. It was not without some complications (a backported function) but should work everywhere out of the box.

Let me know if something needs tweaking.

Thanks!

### CODE REVIEW

**Comment by dnicolodi** [comment]
> I felt that working from standard tooling aligns well with other modes supporting Xref and would be helpful to people used to default keybindings.

Thanks @vkazanov, this is exactly the kind of addition to beancount-mode that I would like to have!

File: beancount-tests.el [hunk 1]
```diff
@@ -333,13 +334,15 @@ known option nmaes."
...
-      (insert "2024-05-11\n")
-      (goto-char 0)
-      (beancount-date-up-day)
-      (should (equal (thing-at-point 'line) "2024-05-12\n"))))
+  :tags '(date-shift)
+  (with-temp-buffer
+    (insert "2024-05-11\n")
+    (goto-char 0)
+    (beancount-date-up-day)
+    (should (equal (thing-at-point 'line) "2024-05-12\n"))))
```
Comment (dnicolodi) [inline beancount-tests.el:345]: I think that this adjustment of the indentation should be applied to the test right below this as well.

File: beancount-tests.el [hunk 2]
```diff
@@ -348,3 +351,95 @@ known option nmaes."
...
       (should (equal (thing-at-point 'line) "2024-05-10\n"))))
+
+;;; Xref backend
+
+(defun beancount-test-xref-definition-pos (identifier position)
+  (let ((defs (xref-backend-definitions 'beancount identifier)))
+    (should (equal (length defs) 1))
+    (let* ((def (car (xref-backend-definitions 'beancount identifier)))
+           (loc (xref-item-location def))
+           ;; Pre-28.1 defclass was used for xref-buffer-location
```
Comment (dnicolodi) [inline beancount-tests.el]: I would spell this a bit differently, for clarity:
```suggestion
           ;; Pre Emacs-28, defclass was used for xref-buffer-location.
```

File: beancount-tests.el [hunk 3]
```diff
@@ -348,3 +351,95 @@ known option nmaes."
       (goto-char 0)
       (beancount-date-down-day)
       (should (equal (thing-at-point 'line) "2024-05-10\n"))))
+
+;;; Xref backend
+
+(defun beancount-test-xref-definition-pos (identifier position)
+  (let ((defs (xref-backend-definitions 'beancount identifier)))
```
Comment (dnicolodi) [inline beancount-tests.el:360]: Can you please add a comment that says what this function does? Not being familiar with the xref implementation, it is not immediately obvious to me

File: beancount-tests.el [hunk 4]
```diff
@@ -348,3 +351,95 @@ known option nmaes."
...
+  Equity:Opening-Balances
+  Assets:Account1        1.00 TDB900
+")
+    (beancount-test-xref-definition-pos "Assets:Account1" 2)
+    (beancount-test-xref-definition-pos "Assets:Account2" 41)
+    (beancount-test-xref-definition-pos "Assets:Account3" 80)))
+
+(ert-deftest beancount/xref-backend-references ()
+  :tags '(xref)
+  (let ((file (make-temp-file "beancount-test-"))
```
Comment (dnicolodi) [inline beancount-tests.el]: Why does this test require a buffer baked by a real file, instead than temporary buffer like the other tests?

File: beancount-tests.el [hunk 5]
```diff
@@ -348,3 +351,95 @@ known option nmaes."
...
+               (should (equal (length (xref-backend-references 'beancount "Assets:Account3")) 1)))
+      (ignore-errors (delete-file file))
+      (ignore-errors
+        (with-current-buffer buf
+          (set-buffer-modified-p nil))
+        (kill-buffer buf)))))
+
+(ert-deftest beancount/xref-backend-apropos ()
+  :tags '(xref)
+  (let ((file (make-temp-file "beancount-test-"))
```
Comment (dnicolodi) [inline beancount-tests.el]: Same for this test.

File: beancount.el [hunk 6]
```diff
@@ -570,30 +578,40 @@ With an argument move to the previous non cleared transaction."
                     (complete-with-action action candidates string pred))))
             (list (match-beginning 1) (match-end 1) completion-table))))))))
 
-(defun beancount-collect (regexp n)
-  "Return an unique list of REGEXP group N in the current buffer."
+(defun beancount-collect-pos-alist (regexp n)
```
Comment (dnicolodi) [inline beancount.el:581]: Can you please move the refactoring to his own commit (and squash the other commits implementing the xref backend into a single commit)?

File: beancount.el [hunk 7]
```diff
@@ -570,30 +578,40 @@ With an argument move to the previous non cleared transaction."
...
           (while (re-search-forward regexp nil t)
             ;; Ignore matches around `pos' (the point position when
             ;; entering this funcyion) since that's presumably what
             ;; we're currently trying to complete.
             (unless (<= (match-beginning 0) pos (match-end 0))
-              (puthash (match-string-no-properties n) nil hash)))
-          (hash-table-keys hash))))))
+              (push (cons (match-string-no-properties n) (match-beginning 0))
+                    result)))
+          (nreverse result))))))
```
Comment (dnicolodi) [inline beancount.el:596]: The reason why `beancount-collect` used an hash table is that the amount of postings in a ledger can be huge, thus accumulating all the account names ever occurring in the ledger easily build ups to a very very long list. Creating a huge list for then immediately dropping most of it in a de-duplication step didn't seem very efficient way, especially because this function is called for completion, thus latency matter. Have you thought about the performance implications?

**Review by dnicolodi — CHANGES_REQUESTED** [review]

File: beancount-tests.el (hunk 4 above)
Comment (vkazanov) [reply beancount-tests.el]: I wanted to do the buffer-backed test but the backend does a call to (buffer-file-name) at some point. While defstruct accepts nil, the older defclasses don't, unfortunately. 

Newer ert (the ert-x.el package) have a macro for this kind of cases but emacs 26.1 doesn't so I came up with this wrapper.

File: beancount-tests.el (hunk 5 above)
Comment (vkazanov) [reply beancount-tests.el]: See above

File: beancount.el (hunk 7 above)
Comment (vkazanov) [reply beancount.el:596]: Yes, and this works for completion as it only needs unique postings. But looking for references means that we have to go through all of the account postings.

So we have to cache unique posting names separately for completion purposes, and build a full list of references separately.

Comment (dnicolodi) [reply beancount.el:596]: I understand why the change is made, no problem with that. I'm just worried that the change may make completion slower for very large ledgers. However, I don't have a very good grasp for elisp performance, thus this may be premature optimization.

One possible solution to keep the optimization and avoid code duplication, would be to make `beancount-collect` get a reference to a function to call for each match found and pass in different functions for the two use cases.

File: beancount-tests.el (hunk 4 above)
Comment (dnicolodi) [reply beancount-tests.el]: It may be time to drop Emacs 26 support. I was looking at integrating the flymake support into `beancount.el` and also that code requires Emacs 27, and I don't want to try to port it to older Emacs.

Comment (dnicolodi) [reply beancount-tests.el]: If we keep Emacs 26 support, and we need this work-around, please add a comment explaining why we need it.

Comment (vkazanov) [reply beancount-tests.el]: I'll create a wrapper macro for this

File: beancount.el (hunk 6 above)
Comment (vkazanov) [reply beancount.el:581]: This might be a bit too entangled by now. But let me try. 

Squashing is easy though :-)

File: beancount-tests.el (hunk 4 above)
Comment (vkazanov) [reply beancount-tests.el]: Added both a wrapper macro and comments

File: beancount.el (hunk 7 above)
Comment (vkazanov) [reply beancount.el:596]: Agree.

I'll have some time today to assess things and do something more efficient.

Comment (vkazanov) [reply beancount.el:596]: @dnicolodi I've just checked a relatively big file (25k lines). On my 2-3 years old notebook there was no noticeable slowdown for both definition search (=M-.=) and inline completion.

This is becase the list is only built once and then a cached list is used. It's the reference search with =M-?= that results in 0.5-1 sec slowdown. 

Let me know if you think speeding things up is still worth the trouble.

File: beancount.el (hunk 6 above)
Comment (vkazanov) [reply beancount.el:581]: I squashed everything into one commit, and my git-fu is not good enough for extracting the refactoring bit into a separate commit. I should have done refactoring separately, sorry, I just went with the flow on this one.

**Comment by vkazanov** [comment]
@dnicolodi, BTW, I have this other idea I want to experiment with: specialize outline-minor-mode's approach a bit to beancount-mode's needs.

The plan:

1. come up with a function that makes it possible to hide or show subsets of the file *based on transaction-related info*. Think outline-minor-mode but smarter than just section delimiters. The buffer is editable, with non-relevant things hidden. I do have a prototype already but it's not ready for prime time yet.
2. a function to only show transactions with a certain tag/link (tags/links would become clickable then).
3. a function to only show transactions within a certain date interval.
3. a function to only transactions related to a certain account.

Given how large the files tend to become this should help quite a bit 

Would you be open to such an improvement?

Obviously, I'd finish the xref backend first.

**Comment by vkazanov** [comment]
@dnicolodi sorry for nudging (gently). Any updates necessary? Thank you

File: beancount.el (hunk 7 above)
Comment (dnicolodi) [reply beancount.el:596]: Sounds good. If it is perceived as slow we can optimize this later.

File: beancount.el (hunk 6 above)
Comment (dnicolodi) [reply beancount.el:581]: I split the commits and pushed to your branch.

File: beancount.el [hunk 8]
```diff
@@ -259,6 +261,11 @@ from the open directive for the relevant account."
 (defconst beancount-metadata-regexp
   "^\\s-+\\([a-z][A-Za-z0-9_-]+:\\)\\s-+\\(.+\\)")
 
+(defconst beancount-open-directive-regexp
+  (concat "^\\(" beancount-date-regexp "\\) +"
+          "\\(open\\) +"
+          "\\(" beancount-account-regexp "\\) +"))
```
Comment (dnicolodi) [inline beancount.el]: This is problematic as it requires the `open` directive to have a space after the account name, but one is not required:
```beancount
2024-06-03 open Assets:Foo
```
is a valid open directive.

**Review by dnicolodi — CHANGES_REQUESTED** [review]

File: beancount.el (hunk 8 above)
Comment (vkazanov) [reply beancount.el]: Ooops. Should be good now.
